pkg/identityserver/rights.go

```diff
@@ -213,3 +214,114 @@ func (is *IdentityServer) UserRights(ctx context.Context, userIDs *ttnpb.UserIde
 	}
 	return entity.Union(universal), nil
 }
+
+var (
+	errInsufficientRights = errors.DefinePermissionDenied(
+		"insufficient_rights",
+		"insufficient rights",
+	)
+	errSomeGatewaysNotFound = errors.DefineNotFound(
+		"some_gateways_not_found",
+		"some gateways not found",
+	)
+)
+
+func (is *IdentityServer) assertGatewayRights( // nolint:gocyclo
+	ctx context.Context,
+	gtwIDs []*ttnpb.GatewayIdentifiers,
+	requiredGatewayRights *ttnpb.Rights,
+) error {
+	authInfo, err := is.authInfo(ctx)
+	if err != nil {
+		return err
+	}
+
+	// If the caller is not an organization or user, there's nothing more to do.
+	ouID := authInfo.GetOrganizationOrUserIdentifiers()
+	if ouID == nil {
+		return errInsufficientRights.New()
+	}
+
+	// Check that the caller has the requested rights.
+	authInfoRights := ttnpb.RightsFrom(authInfo.GetRights()...)
+	if !authInfoRights.IncludesAll(requiredGatewayRights.GetRights()...) {
+		return errInsufficientRights.New()
+	}
+
+	return is.store.Transact(ctx, func(ctx context.Context, st store.Store) error {
+		gtws, err := st.FindGateways(ctx, gtwIDs, []string{"ids", "status_public", "location_public"})
+		if err != nil {
+			return err
+		}
+		if len(gtws) != len(gtwIDs) {
+			if is.IsAdmin(ctx) {
+				// Return the cause only to the admin.
+				// This follows the same logic as in ListRights.
+				return errSomeGatewaysNotFound.New()
+			}
+			return errInsufficientRights.New()
+		}
+
+		// Filter out the public gateways from membership checks,
+		// when only public fields are requested.
+		publicGatewayIds := make(map[string]struct{}, len(gtws))
+		bothStatsAndLocation := len(requiredGatewayRights.Sub(
+			ttnpb.RightsFrom(
+				ttnpb.Right_RIGHT_GATEWAY_LOCATION_READ,
+				ttnpb.Right_RIGHT_GATEWAY_STATUS_READ,
+			),
+		).GetRights()) == 0
+		onlyPublicLocation := len(requiredGatewayRights.Sub(
+			ttnpb.RightsFrom(
+				ttnpb.Right_RIGHT_GATEWAY_LOCATION_READ,
+			),
+		).GetRights()) == 0
+		onlyPublicStats := len(requiredGatewayRights.Sub(
+			ttnpb.RightsFrom(
+				ttnpb.Right_RIGHT_GATEWAY_LOCATION_READ,
```

Comment:
Is this right correct ? Seems like `RIGHT_GATEWAY_LOCATION_READ` is referenced twice.